server/config.py

```diff
@@ -26,23 +28,40 @@
 KEY_SECURE_SESSION_COOKIES = 'secure_session_cookies'
 KEY_SAMESITE_SESSION_COOKIES = 'samesite_session_cookies'
 
+from distutils.util import strtobool
 
 #
 # Container for properties that should be read on startup, and maybe be
 # refreshed dynamically.
 #
-class Config:
+# If using redis all the keys are cached in the redis server, otherwise
+# the keys are just the attributes of this class.  Note that redis does
+# not support boolean values.  Hence care must be taken with the string
+# conversion.
+#
+class Config(dict):
+
+    def __init__(self):
+        if constants.REDIS_IP_ADDR is not None:
+            self.red = redis.Redis(constants.REDIS_IP_ADDR, port=6379)
+
+    def __getitem__(self, key):
+        if constants.REDIS_IP_ADDR is not None:
+            # TODO: Rethink how this is done as it will break if we store anything in the config other than a boolean.
+            return True if strtobool(str(self.red.get(key), 'utf-8')) else False
+        else:
+            return super(Config, self).__getitem__(key)
 
-    training_enabled = True
-    use_tpu = True
-    secure_session_cookies = True
-    samesite_session_cookies = True
+    def __setitem__(self, key, value):
+       if constants.REDIS_IP_ADDR is not None:
+           self.red.set(key, str(value))
+       super(Config, self).__setitem__(key, value)
 
     def reset(self):
-        self.training_enabled = True
-        self.use_tpu = True
-        self.secure_session_cookies = True
-        self.samesite_session_cookies = True
+        self[KEY_TRAINING_ENABLED] = True
+        self[KEY_USE_TPU] = True
+        self[KEY_SECURE_SESSION_COOKIES] = True
+        self[KEY_SAMESITE_SESSION_COOKIES] = True
 
     def refresh(self):
```

Comment:
Add some explanation... If app engine instance 2 calls refresh, it will read from the datastore and cache it in redis. App engine instance 1 will get updated values from redis, although its dictionary will no longer match the values in redis.
As we discussed, this does not lead to a problem because we don't actually use the dictionary if redis is enabled.

Reply:
I've added explanatory text up in the larger comment block by the class declaration.